WIP: New Python SDK (#251)
Unstructured-IO/docs: api-reference/api-services/sdk-python.mdx

    <Note>
        The SDK uses semantic versioning and major bumps could bring breaking changes. It is advised to

>         pin your installed version. See the [migration guide](#migration-guide) below for breaking change announcements.

Reviewer: ```suggestion
        pin your installed version. See the [migration guide](#migration-guide), later on this page, for breaking change announcements.
```

    </Note>

## Basics
    Let's start with a simple example in which you send a PDF document to be partitioned with the Unstructured API. Note that the Python SDK gives you a choice of two interfaces. The functionality is identical, but the Pydantic model syntax may provide completion hints in your IDE, for instance.

    <CodeGroup>
    ```python Python (TypedDict example)
    import os, json

    import unstructured_client
    from unstructured_client.models import shared

    client = unstructured_client.UnstructuredClient(
        api_key_auth=os.getenv("UNSTRUCTURED_API_KEY"),
        server_url=os.getenv("UNSTRUCTURED_API_URL"),
    )

    filename = "PATH_TO_INPUT_FILE"

    req = {
        "partition_parameters": {
            "files": {
                "content": open(filename, "rb"),
                "file_name": filename,
            },
            "strategy": shared.Strategy.HI_RES,
            "languages": ['eng'],

>             "split_pdf_page": True,

Reviewer: Should this also have the following, like our other examples?

```
"split_pdf_allow_failed": True,
"split_pdf_concurrency_level": 15
```

Author: Will add these per the other comment

        pin your installed version.
    </Note>

> ### Notes on breaking changes in 0.26.0

Reviewer: This should be moved to the end and titled something like "Migration guide." Then a note here like "if you are upgrading from Python SDK version 0.25.0 or earlier to version 0.26.0 or later, see the <link to migration guide> for breaking changes that will impact your code." Leaving all of this here, especially for new users, seems too visually cluttering and distracting.

    )
)

resp = s.general.partition(request=req)
```

- If you see the error: `TypeError: BaseModel.__init__() takes 1 positional argument but 2 were given`

The PartitionRequest constructor no longer allows for positional arguments. You must specify partition_parameters by name.

```python
# Instead of:
req = operations.PartitionRequest(
    shared.PartitionParameters(
        files=files,
    )
)

# Switch to:
req = operations.PartitionRequest(
    partition_parameters=shared.PartitionParameters(
        files=files,
    )
)
```

- If you see the error: `TypeError: General.partition() takes 1 positional argument but 2 were given`

Likewise, the partition function no longer allows for positional arguments. You must specify request by name.

```python
# Instead of:
resp = s.general.partition(req)

# Switch to:
resp = s.general.partition(request=req)
```

## Basics
    Let's start with a simple example in which you send a PDF document to be partitioned with the Unstructured API. Note that the Python SDK gives you a choice of two interfaces. The functionality is identical, but the Pydantic model syntax may provide completion hints in your IDE, for instance.

    <CodeGroup>

>     ```python Python (TypedDict example)

Reviewer: It sounds like we don't have a recommendation over which one that customers should use. That being said, should we choose one and use that throughout the docs whenever we show usage examples for the Python SDK. If so, which one? And that one should come first or left-most in the `CodeGroup`.

Reviewer: (And we should probably update all of the Python SDK examples across the docs to now show how to use the multi-file async pattern only?)

Reviewer: (A quick follow-up: we can wait until later to update all of the Python SDK examples across the docs to now show how to use the multi-file async pattern only.)

    )

    filename = "PATH_TO_INPUT_FILE"
    with open(filename, "rb") as f:
        data = f.read()

    req = operations.PartitionRequest(
        partition_parameters=shared.PartitionParameters(
            files=shared.Files(
                content=data,
                content=open(filename, "rb"),
                file_name=filename,
            ),
            strategy=shared.Strategy.HI_RES,
            languages=['eng'],
            split_pdf_page=True,            # If True, splits the PDF file into smaller chunks of pages.
            split_pdf_allow_failed=True,    # If True, the partitioning continues even if some pages fail.

>             split_pdf_concurrency_level=15  # Set the number of concurrent request to the maximum value: 15.

Reviewer: Why remove these?

    except Exception as e:
        print(e)
    ```
    </CodeGroup>

## Async Partitioning

    The Python SDK also has a `partition_async`. This call is equivalent to `partition` except that it can be used in a non blocking context. For instance, `asyncio.gather` can be used to concurrently process multiple files at once, as in this example: 

    <CodeGroup>
    ```python
    import asyncio
    import os, json

    import unstructured_client
    from unstructured_client.models import shared

    client = unstructured_client.UnstructuredClient(
        api_key_auth=os.getenv("UNSTRUCTURED_API_KEY"),
        server_url=os.getenv("UNSTRUCTURED_API_URL"),
    )

>     async def call_api(filename):

Reviewer: If I'm reading this correctly, shouldn't the signature be something like `call_api(infile, outfile)`?

Let's say I call this three times, with `PATH_TO_INPUT_FILE_1`, `PATH_TO_INPUT_FILE_2`, and `PATH_TO_INPUT_FILE_3`, won't `PATH_TO_OUTPUT_FILE` only show the results of processing   `PATH_TO_INPUT_FILE_3` as a result of the third and final call to `call_api`?

Author: I changed it to reuse the passed in filename plus `.json`

        req = {
            "partition_parameters": {
                "files": {
                    "content": open(filename, "rb"),
                    "file_name": filename,
                },
                "strategy": shared.Strategy.HI_RES,
            }
        }

        try:
            res = await client.general.partition_async(request=req)

            element_dicts = [element for element in res.elements]

            # Write the processed data to a local file.
            json_elements = json.dumps(element_dicts, indent=2)

            # Each input will need its own output file
            with open("PATH_TO_OUTPUT_FILE", "w") as file:
                file.write(json_elements)

        except Exception as e:
            print(e)

    async def process_files(filenames):

>         filenames = [

Reviewer: What if I have a folder with more than 3 files...is there any way to specify wildcards or folder/file globbing to say "process all the files in the root as well as recursively throughout this folder's hierarchy?"

Author: I'm mostly concerned with documenting the breaking changes for the new 0.26.0 here. Once this is out and the new SDK is published, let's revisit the usage examples.

Reviewer: Then we would only want to show one file here instead of three, right?

Reviewer: Also, how should we re-word this related usage note?

![Screenshot 2024-09-30 at 1 58 00 PM](https://github.com/user-attachments/assets/1347329d-959f-4b6e-8a8c-b47856c84823)

Author: Hmm, yeah. This will give us parity with being able to run multiple documents at once. The specific advantages of ingest would be:
- Prebuilt connectors (i.e. the cloud storage bullet)
- Batch processing of a whole directory, with caching of the results. Now you'll be able to code up a directory scraper using the SDK, but ingest will be an out of the box experience.
- Pipeline management. Ingest can extract and chunk separately, etc. This is much nicer than rolling your own in the SDK

Author: Let me take a stab at rewording that

Reviewer: How about:

> - You need to work with documents in cloud storage.
> - You want to cache the results of processing multiple files in batches.
> - You want more precise control over document-processing pipeline stages such as partitioning, chunking, filtering, staging, and embedding.

Author: Perfect, just changed!

    except Exception as e:
        print(e)
    ```
    </CodeGroup>

> ## Async Partitioning

Reviewer: ```suggestion
## Async partitioning
```

Reviewer: It seems to me that `partition_async` would typically be more flexible and robust than `partition`. If so, shouldn't we present the `partition_async` content before the `partition` content?

    The Python SDK also has a `partition_async`. This call is equivalent to `partition` except that it can be used in a non blocking context. For instance, `asyncio.gather` can be used to concurrently process multiple files at once, as demonstrated here: 

    <CodeGroup>
    ```python
    import asyncio
    import os, json

    import unstructured_client
    from unstructured_client.models import shared

    client = unstructured_client.UnstructuredClient(
        api_key_auth=os.getenv("UNSTRUCTURED_API_KEY"),
        server_url=os.getenv("UNSTRUCTURED_API_URL"),
    )

    async def call_api(filename):
        req = {
            "partition_parameters": {
                "files": {
                    "content": open(filename, "rb"),
                    "file_name": filename,
                },
                "strategy": shared.Strategy.HI_RES,
            }
        }

        try:
            res = await client.general.partition_async(request=req)

            element_dicts = [element for element in res.elements]

            json_elements = json.dumps(element_dicts, indent=2)

>             output_filename = filename + ".json"  # Save off the json response alongside the input file

Reviewer: ```suggestion
            output_filename = filename + ".json"  # Save the JSON response alongside the input file.
```

* Refer to the [API parameters](/api-reference/api-services/api-parameters) page for the full list of available parameters.
* Refer to the [Examples](/api-reference/api-services/examples) page for some inspiration on using the parameters.

## Migration guide

> ### Upgrading to 0.26.0

Reviewer: Minor nit: this is probably fine, as we will likely add more post-0.26.0 migration sections later. But having a section with only one sub-section is a tiny bit visually jarring.

- If you see the error: `AttributeError: 'PartitionParameters' object has no attribute 'partition_parameters'`

> Previously, the SDK has accepted a PartitionParameters object as input to the sdk.general.partition function. Now, this object must be wrapped in a PartitionRequest object. The old behavior was deprecated in 0.23.0 and removed in 0.26.0.

Reviewer: Minor nit: it's a good practice to line up everything "underneath" or "within" a bullet for better readability. This means that each and every line under the bullet must be indented 2 spaces. Alternatively, use `###` or bold instead.

> Previously, the SDK has accepted a PartitionParameters object as input to the sdk.general.partition function. Now, this object must be wrapped in a PartitionRequest object. The old behavior was deprecated in 0.23.0 and removed in 0.26.0.

Reviewer: ```suggestion
Previously, the SDK accepted a `PartitionParameters` object as input to the `sdk.general.partition` function. Now, this object must be wrapped in a `PartitionRequest` object. The old behavior was deprecated in 0.23.0 and removed in 0.26.0.
```

```python
# Instead of:
from unstructured_client.models import shared

req = shared.PartitionParameters(
    files=files,
)

resp = s.general.partition(request=req)

# Switch to:
from unstructured_client.models import shared, operations

req = operations.PartitionRequest(
    partition_parameters=shared.PartitionParameters(
        files=files,
    )
)

resp = s.general.partition(request=req)
```

- If you see the error: `TypeError: BaseModel.__init__() takes 1 positional argument but 2 were given`

> In 0.26.0 the PartitionRequest constructor no longer allows for positional arguments. You must specify partition_parameters by name.

Reviewer: ```suggestion
In 0.26.0, the `PartitionRequest` constructor no longer allows for positional arguments. You must specify `partition_parameters` by name.
```

```python
# Instead of:
req = operations.PartitionRequest(
    shared.PartitionParameters(
        files=files,
    )
)

# Switch to:
req = operations.PartitionRequest(
    partition_parameters=shared.PartitionParameters(
        files=files,
    )
)
```

- If you see the error: `TypeError: General.partition() takes 1 positional argument but 2 were given`

> Likewise, in 0.26.0 the partition function no longer allows for positional arguments. You must specify request by name.

Reviewer: ```suggestion
In 0.26.0, the `partition` function no longer allows for positional arguments. You must specify `request` by name.
```
